src/connection.rs

```diff
@@ -8,793 +8,857 @@
+            config: Arc::new(cfg),
+            socket,
+            streams: IntMap::default(),
+            control_sender,
+            control_receiver: Pausable::new(control_receiver),
+            stream_sender,
+            stream_receiver,
             next_id: match mode {
                 Mode::Client => 1,
                 Mode::Server => 2
-            }
+            },
+            garbage: Vec::new(),
+            shutdown: Shutdown::NotStarted
         }
     }
 
-    fn next_stream_id(&mut self) -> Result<stream::Id, ConnectionError> {
-        let proposed = stream::Id::new(self.next_id);
-        self.next_id = self.next_id.checked_add(2).ok_or(ConnectionError::NoMoreStreamIds)?;
-        match self.mode {
-            Mode::Client => assert!(proposed.is_client()),
-            Mode::Server => assert!(proposed.is_server())
-        }
-        Ok(proposed)
+    /// Get a controller for this connection.
+    pub fn control(&self) -> Control {
+        Control::new(self.control_sender.clone())
     }
 
-    fn is_valid_remote_id(&self, id: stream::Id, ty: Type) -> bool {
-        match ty {
-            Type::Ping | Type::GoAway => return id.is_session(),
-            _ => {}
+    /// Get the next incoming stream, opened by the remote.
+    ///
+    /// This must be called repeatedly in order to make progress.
+    /// Once `Ok(None)` or `Err(_)` is returned the connection is
+    /// considered closed and no further invocation of this method
+    /// must be attempted.
+    ///
+    /// # Cancellation
+    ///
+    /// Please note that if you poll the returned [`Future`] it *must
+    /// not be cancelled* but polled until [`Poll::Ready`] is returned.
+    pub async fn next_stream(&mut self) -> Result<Option<Stream>> {
+        let result = self.next().await;
+
+        if let Ok(Some(_)) = result {
+            return result
         }
-        match self.mode {
-            Mode::Client => id.is_server(),
-            Mode::Server => id.is_client()
+
+        // At this point we are either at EOF or encountered an error.
+        // We close all streams and wake up the associated tasks before
+        // closing the socket. The connection is then considered closed.
+
+        self.drop_all_streams();
+
+        if let Err(ConnectionError::Closed) = result {
+            return Ok(None)
         }
+
+        result
     }
 
-    fn add_pending(&mut self, frame: RawFrame) -> Result<(), ConnectionError> {
-        if self.pending.len() >= self.config.max_pending_frames {
-            return Err(ConnectionError::TooManyPendingFrames)
+    /// Get the next inbound `Stream` and make progress along the way.
+    ///
+    /// This is called from `Connection::next_stream` instead of being a
+    /// public method itself in order to guarantee proper closing in
+    /// case of an error or at EOF.
+    async fn next(&mut self) -> Result<Option<Stream>> {
+        loop {
+            self.garbage_collect().await?;
```

Comment:
Can this not happen outside of the loop? Can the `strong_count` of a stream go down within the loop (I didn't see how)?

Reply:
Ah, I guess streams can be dropped between yield points (i.e. `.await`).

Reply:
The strong count of a `Stream`'s `Arc` can change at any time by just dropping the `Stream`. If we were to garbage collect only once outside the loop, we risk adding many new (outbound) streams (depending on the number of control commands) while not removing dropped ones.